prevent running script in wrong account

GavinFrazar opened the pull request:
Part of https://github.com/gravitational/teleport/issues/44367
- https://github.com/gravitational/teleport/issues/44367

This PR catches the case where the user mistakenly opens cloud shell in a different AWS account than the integration. Now the script will catch that mistake early instead of giving a cryptic "role not found" error.

marcoandredinis commented on lib/web/integrations_awsoidc.go:
```diff
@@ -270,11 +270,15 @@ func (h *Handler) awsOIDCConfigureDeployServiceIAM(w http.ResponseWriter, r *htt
 
 	// Ensure the IntegrationName is valid.
-	_, err = h.GetProxyClient().GetIntegration(ctx, integrationName)
+	integration, err := h.GetProxyClient().GetIntegration(ctx, integrationName)
 	// NotFound error is ignored to prevent disclosure of whether the integration exists in a public/no-auth endpoint.
 	if err != nil && !trace.IsNotFound(err) {
 		return nil, trace.Wrap(err)
 	}
+	parsed, err := aws.ParseRoleARN(integration.GetAWSOIDCIntegrationSpec().RoleARN)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```
The integration might not exist, and in that case `integration` will be `nil`, thus we can't get the RoleARN.
For those scenarios, we could return clearly fake account id (`000000000000`).
However, that would disclose that this integration does not exist in a public/no-auth endpoint.

GavinFrazar replied:
thanks for catching this, I didn't notice the not found error handling above.

I've reworked the PR to pass the account id via query param as you suggested, so the point is moot now.

marcoandredinis commented on lib/web/integrations_awsoidc.go:
```diff
@@ -300,6 +304,7 @@ func (h *Handler) awsOIDCConfigureDeployServiceIAM(w http.ResponseWriter, r *htt
 		fmt.Sprintf("--aws-region=%s", shsprintf.EscapeDefaultContext(awsRegion)),
 		fmt.Sprintf("--role=%s", shsprintf.EscapeDefaultContext(role)),
 		fmt.Sprintf("--task-role=%s", shsprintf.EscapeDefaultContext(taskRole)),
+		fmt.Sprintf("--aws-account=%s", shsprintf.EscapeDefaultContext(parsed.AccountID)),
```
If some bad actor guesses the integration name, they will get access to the account id as well.
AccountID is not considered a secret, but being able to get it from a public/no-auth endpoint might raise some security concerns.

GavinFrazar replied:
addressed by passing the account id as a query param

marcoandredinis commented on tool/teleport/common/teleport.go:
```diff
@@ -472,6 +472,7 @@ func Run(options Options) (app *kingpin.Application, executedCommand string, con
 	integrationConfDeployServiceCmd.Flag("aws-region", "AWS Region.").Required().StringVar(&ccf.IntegrationConfDeployServiceIAMArguments.Region)
 	integrationConfDeployServiceCmd.Flag("role", "The AWS Role used by the AWS OIDC Integration.").Required().StringVar(&ccf.IntegrationConfDeployServiceIAMArguments.Role)
 	integrationConfDeployServiceCmd.Flag("task-role", "The AWS Role to be used by the deployed service.").Required().StringVar(&ccf.IntegrationConfDeployServiceIAMArguments.TaskRole)
+	integrationConfDeployServiceCmd.Flag("aws-account", "The AWS account ID.").Required().StringVar(&ccf.IntegrationConfDeployServiceIAMArguments.AccountID)
```
```suggestion
	integrationConfDeployServiceCmd.Flag("aws-account-id", "The AWS account ID.").Required().StringVar(&ccf.IntegrationConfDeployServiceIAMArguments.AccountID)
```
For consistency with
https://github.com/gravitational/teleport/blob/1531595d12e68139d5432af90a5dee5ada4e5324/tool/teleport/common/teleport.go#L286
and
https://github.com/gravitational/teleport/blob/1531595d12e68139d5432af90a5dee5ada4e5324/tool/teleport/common/teleport.go#L241

marcoandredinis reviewed (commented):
Can we instead send the AccountID from the UI, adding it as a query param when building the URL for the script?
Exposing it in a public/no-auth endpoint might raise security issues.

marcoandredinis commented:
Also, this change should be applied to all the aws oidc configuration scripts (except for the very first one, where we create the IdP in AWS).

GavinFrazar commented:
I've updated the PR to affect all of our scripts and to pass the aws account id via query param from the web ui: https://github.com/gravitational/teleport/compare/5043007539641cb9b144a91aa3616daeaaee5a0f..0dd98e2e3fdb1b8c5349597c8ff9ec75eda5ea32

I need to update the web ui in TAG as well for the `access-graph-cloud-sync-iam.sh` script, so I'll raise a PR in the access-graph repo for that.

marcoandredinis commented:
Given that TAG deployments happen at different times, we have to ensure that

A) TAG UI always sends the account ID and current releases of teleport do not fail on additional arguments

B) If the account id argument is not there (in the `teleport integration configure aws-sync --aws-account-id`), the script must not return an error.

I'm not sure we need to implement both, but we should check with the TAG team to get their feedback on this.

GavinFrazar commented:
> Given that TAG deployments happen at different times, we have to ensure that
> 
> A) TAG UI always sends the account ID and current releases of teleport do not fail on additional arguments
> 
> B) If the account id argument is not there (in the `teleport integration configure aws-sync --aws-account-id`), the script must not return an error.
> 
> I'm not sure we need to implement both, but we should check with the TAG team to get their feedback on this.

I think the easiest way to avoid compatibility issues for these scripts is to just make the `--aws-account-id` optional instead of required. If it's given, we will check the account matches, otherwise it's just like the old behavior that doesn't check.

I'll update the PR to do that.

GavinFrazar commented:
> A) TAG UI always sends the account ID and current releases of teleport do not fail on additional arguments

I forgot to mention how this is addressed: in the TAG PR I have the account id passed as a query param to the teleport web ui.
So in TAG, it tells the user to run a command like this:
```
bash -c "curl 'https://gavin-leaf.cloud.gravitational.io/webapi/scripts/integrations/configure/access-graph-cloud-sync-iam.sh?kind=aws-iam&role=alpha-devteleport-com&awsAccountID=123123123123'"
```

If Teleport proxy recognizes that query param, then the script it generates will include `--aws-account-id`.
Thus, only customers who have updated their teleport version to one that includes this PR's change will actually run a script that passes `--aws-account-id`.

By making `--aws-account-id` optional, your other concern is also addressed.

GavinFrazar committed:
prevent running script in wrong account

All OIDC configuration scripts that we tell users to run in AWS
CloudShell will now check that the script is run in the same AWS
account as the AWS OIDC integration.

GavinFrazar commented:
force pushed to drop the changes in ./api that exported a `ParseRoleARN` func, since that was no longer needed after reworking the PR

r0mant reviewed (approved):
Bot.
